grafana/grafana.libsonnet

```diff
@@ -42,13 +43,15 @@ local k = import 'ksonnet/ksonnet.beta.3/k.libsonnet';
       secret.mixin.metadata.withNamespace($._config.namespace),
     dashboardDefinitions:
       local configMap = k.core.v1.configMap;
-      [
-        local dashboardName = 'grafana-dashboard-' + std.strReplace(name, '.json', '');
-        configMap.new(dashboardName, { [name]: std.manifestJsonEx($._config.grafana.dashboards[name], '    ') }) +
-        configMap.mixin.metadata.withNamespace($._config.namespace)
+      configMapList.new(
```

Comment:
see here, it used to be a `[]`, but now a single `{ kind: ConfigmapList, items: [] }` object, it allows embedding it

Reply:
right, but when you render out a `kind: List`, does kubectl actually create deep lists recursively?

Reply:
Doesn't seem so. See same transformation in kube-prometheus:

https://github.com/coreos/kube-prometheus/blob/3fbc968930c2fef025470c59060d25635070f8e7/jsonnet/kube-prometheus/kube-prometheus.libsonnet#L18

Reply:
That only works because what is generated is not a List object but a directory of objects, where one of the files happens to be a ConfigMapList, so for kubectl the ConfigMapList is actually the top level item.

Reply:
let me run few tests, I'll come back  to you

Reply:
objects side-by-side:

```
$ kubectl create configmap t1 --from-literal=toplevel1=abc --dry-run -o json > /tmp/tl1.json
$ kubectl create configmap t2 --from-literal=toplevel2=abc --dry-run -o json > /tmp/tl2.json

$ cat /tmp/tl1.json /tmp/tl2.json | kubectl -n default create -f  -
configmap/t1 created
configmap/t2 created

$ kubectl -n default get configmap -o json > /tmp/tlist.json

$kubectl -n default delete configmap t1 t2
configmap "t1" deleted
configmap "t2" deleted

# now the real test, 2 objects side by side, one ConfigMap , another ConfigMapList
$ cat /tmp/tlist.json /tmp/tl3.json | tee /dev/stderr | kubectl -n default create -f -
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "v1",
            "data": {
                "toplevel1": "abc"
            },
            "kind": "ConfigMap",
            "metadata": {
                "creationTimestamp": "2019-05-28T14:30:43Z",
                "name": "t1",
                "namespace": "default",
                "resourceVersion": "146855860",
                "selfLink": "/api/v1/namespaces/default/configmaps/t1",
                "uid": "2caf28bd-8155-11e9-bacb-005056bced2e"
            }
        },
        {
            "apiVersion": "v1",
            "data": {
                "toplevel2": "abc"
            },
            "kind": "ConfigMap",
            "metadata": {
                "creationTimestamp": "2019-05-28T14:30:43Z",
                "name": "t2",
                "namespace": "default",
                "resourceVersion": "146855862",
                "selfLink": "/api/v1/namespaces/default/configmaps/t2",
                "uid": "2cb13d6e-8155-11e9-9309-005056bcaf32"
            }
        }
    ],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
{
    "kind": "ConfigMap",
    "apiVersion": "v1",
    "metadata": {
        "name": "t3",
        "creationTimestamp": null
    },
    "data": {
        "toplevel2": "abc"
    }
}
configmap/t1 created
configmap/t2 created
configmap/t3 created

```

Reply:
Right, but try to put the two ConfigMaps into a ConfigMapList within the List object.

Reply:
why would we have this situation?

Reply:
correct :)

Reply:
"library" object should still have "standard" shape:

```
{ key: { kind: ... } }
```

to make it easier to use it in other jsonnet deployments.

it is only examples which need to change. I can go with

```
local grafana =
  ((import 'grafana/grafana.libsonnet') + .... ).grafana;

local flatten(lst) = std.foldl(function(acc, r) if ! std.endsWith(r.kind, "List") then acc + [r] else acc + r.items, lst, []);

k.core.v1.list.new(flatten(grafana[k] for k in grafana))
```

what do you think?

Reply:
yeah, I quite like that

Reply:
Another approach would be to return plain array and change exampes to call `jsonnet --yaml-stream`. 

Either way output is pipeable into kubectl . Which  one do you prefer?

Reply:
I think I prefer the first, as that way when people want to they can still make use of the *List types for grouping, in the way I think it's good that the dashboard configmaps are currently grouped that way.